packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php

```diff
@@ -0,0 +1,209 @@
+                continue;
+            }
+
+            if (! $this->isInAllowedNamespace($allowedNamespacePatterns, $classLikeName)) {
+                $nativeReflectionClass = $classReflection->getNativeReflection();
+                $errorMessage = sprintf(
+                    self::ERROR_MESSAGE,
+                    $classLikeName,
+                    Json::encode($allowedNamespacePatterns, JSON_THROW_ON_ERROR),
+                    $nativeReflectionClass->getNamespaceName(),
+                );
+
+                return [$errorMessage];
+            }
+        }
+
+        return [];
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Define in which namespaces (using *, ** or ? glob-like pattern matching) can classes extending specified class or implementing specified interface exist', [
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+namespace App;
+
+// AbstractType implements \Symfony\Component\Form\FormTypeInterface
+use Symfony\Component\Form\AbstractType;
+
+class UserForm extends AbstractType
+{
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+namespace App\Form;
+
+use Symfony\Component\Form\AbstractType;
+
+class UserForm extends AbstractType
+{
+}
+CODE_SAMPLE
+                ,
+                [
+                    'guards' => [
+                        'Symfony\Component\Form\FormTypeInterface' => ['App\Form\**'],
+                    ],
+                ]
+            ),
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+namespace App\Services;
+
+use App\Component\PriceEngine\PriceProviderInterface;
+
+class CustomerProductProvider extends PriceProviderInterface
+{
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+namespace App\Component\PriceEngineImpl;
+
+use App\Component\PriceEngine\PriceProviderInterface;
+
+class CustomerProductProvider extends PriceProviderInterface
+{
+}
+CODE_SAMPLE
+                ,
+                [
+                    'guards' => [
+                        'App\Component\PriceEngine\**' => [
+                            'App\Component\PriceEngine\**',
+                            'App\Component\PriceEngineImpl\**',
+                        ],
+                    ],
+                ]
+            ),
+        ]);
+    }
+
+    private function isSubjectToGuardedTypeOrNamespacePattern(
+        ClassReflection $classReflection,
+        string $guardedTypeOrNamespacePattern,
+    ): bool {
+        $isGuardedSubjectNamespacePattern = str_contains($guardedTypeOrNamespacePattern, '*') || str_contains(
+            $guardedTypeOrNamespacePattern,
+            '?'
+        );
+        $isGuardedSubjectType = ! $isGuardedSubjectNamespacePattern;
+
+        if ($isGuardedSubjectType && ! $classReflection->isSubclassOf($guardedTypeOrNamespacePattern)) {
+            return false;
+        }
+
+        if ($isGuardedSubjectNamespacePattern) {
```

Comment:
you  should can flip check by if (! $isGuardedSubjectNamespacePattern) and return true early, so next you can directly return $isGuardedSubjectSubclassOfGuardedPattern